rest/oidc_api_test.go

```diff
@@ -1,139 +1,816 @@
+			mockAuthServer.options.issuer = tc.issuer
+
+			// Check _oidc_challenge behavior
+			authURL := strings.ReplaceAll(tc.authURL, "${path}", "_oidc_challenge")
+			resp := rt.SendAdminRequest(http.MethodGet, authURL, "")
+			require.Equal(t, http.StatusUnauthorized, resp.Code)
+			wwwAuthHeader := resp.Header().Get("Www-Authenticate")
+			location := regexp.MustCompile(`login="(?P<login>.*?)"`).FindStringSubmatch(wwwAuthHeader)[1]
+
+			require.NotEmpty(t, location, "error extracting location from header")
+			locationURL, err := url.Parse(location)
+			require.NoError(t, err, "error parsing location URL")
+			redirectURI := locationURL.Query().Get(requestParamRedirectURI)
+			redirectURL, err := url.Parse(redirectURI)
+			require.NoError(t, err, "error parsing redirect_uri URL")
+			assert.Equal(t, tc.wantProvider, redirectURL.Query().Get(auth.OIDCAuthProvider))
+
+			// Check _oidc behavior
+			authURL = strings.ReplaceAll(tc.authURL, "${path}", "_oidc")
+			resp = rt.SendAdminRequest(http.MethodGet, authURL, "")
+			require.Equal(t, http.StatusFound, resp.Code)
+			location = resp.Header().Get("Location")
+
+			require.NotEmpty(t, location, "error extracting location from header")
+			locationURL, err = url.Parse(location)
+			require.NoError(t, err, "error parsing location URL")
+			redirectURI = locationURL.Query().Get(requestParamRedirectURI)
+			redirectURL, err = url.Parse(redirectURI)
+			require.NoError(t, err, "error parsing redirect_uri URL")
+			assert.Equal(t, tc.wantProvider, redirectURL.Query().Get(auth.OIDCAuthProvider))
+		})
+	}
+}
+
+// Checks End to end OpenID Connect Authorization Code flow.
+func TestOpenIDConnectAuth(t *testing.T) {
+	type test struct {
+		name            string
+		providers       auth.OIDCProviderMap
+		defaultProvider string
+		authURL         string
+		wantError       wantError
+	}
+	tests := []test{
+		{
+			// Successful new user authentication when auto registration is enabled and a single provider
+			// is configured through providers configuration. Explicitly specified to include access token in
+			// token response.
+			name: "successful user registration against single provider",
+			providers: auth.OIDCProviderMap{
+				"foo": &auth.OIDCProvider{
+					Name:               "foo",
+					ClientID:           base.StringPointer("baz"),
+					UserPrefix:         "foo",
+					ValidationKey:      base.StringPointer("qux"),
+					Register:           true,
+					IncludeAccessToken: true,
+				},
+			},
+			defaultProvider: "foo",
+			authURL:         "/db/_oidc?provider=foo&offline=true",
+		}, {
+			// Unsuccessful new user authentication when auto registration is enabled and a single provider
+			// is configured through providers configuration. Explicitly specified to include access token in
+			// token response.
+			name: "unsuccessful user registration against single provider",
+			providers: auth.OIDCProviderMap{
+				"foo": &auth.OIDCProvider{
```

Comment:
A lot of these tests use the same generic provider.  Defining this generic provider ('fooProvider') once prior to the test cases would do two things: reduce the amount of cloned code, but more importantly make it more obvious when a given test is doing something different with the provider.